src/python/pants/backend/python/rules/pytest_coverage.py

```diff
@@ -0,0 +1,353 @@
+        [FileContent(path=".coveragerc", content=coveragerc_content.encode()),]
+    )
+
+
+def get_coverage_plugin_input():
+    return InputFilesContent(
+        FilesContent(
+            (
+                FileContent(
+                    path=f"{COVERAGE_PLUGIN_MODULE_NAME}.py",
+                    content=pkg_resources.resource_string(__name__, "coverage_plugin/plugin.py"),
+                ),
+            )
+        )
+    )
+
+
+def get_packages_to_cover(
+    *, target: PythonTestsAdaptor, source_root_stripped_file_paths: Tuple[str, ...],
+) -> Tuple[str, ...]:
+    if hasattr(target, "coverage"):
+        return tuple(sorted(set(target.coverage)))
+    return tuple(
+        sorted(
+            {
+                os.path.dirname(source_root_stripped_source_file_path).replace(
+                    os.sep, "."
+                )  # Turn file paths into package names.
+                for source_root_stripped_source_file_path in source_root_stripped_file_paths
+            }
+        )
+    )
+
+
+def ensure_section(config_parser: configparser.ConfigParser, section: str) -> None:
+    """Ensure a section exists in a ConfigParser."""
+    if not config_parser.has_section(section):
+        config_parser.add_section(section)
+
+
+def construct_coverage_config(
+    source_roots: SourceRoots, python_files: List[str], test_time: Optional[bool] = False,
+) -> str:
+    # A map from source root stripped source to its source root. eg:
+    #  {'pants/testutil/subsystem/util.py': 'src/python'}
+    # This is so coverage reports referencing /chroot/path/pants/testutil/subsystem/util.py can be mapped
+    # back to the actual sources they reference when merging coverage reports.
+    init_files = list(identify_missing_init_files(list(python_files)))
+
+    def source_root_stripped_source_and_source_root(file_name):
+        source_root = source_roots.find_by_path(file_name)
+        source_root_stripped_path = file_name[len(source_root.path) + 1 :]
+        return (source_root_stripped_path, source_root.path)
+
+    source_to_target_base = dict(
+        source_root_stripped_source_and_source_root(filename)
+        for filename in sorted(python_files) + init_files
+    )
+    config_parser = configparser.ConfigParser()
+    config_parser.read_file(StringIO(DEFAULT_COVERAGE_CONFIG))
+    ensure_section(config_parser, "run")
+    config_parser.set("run", "plugins", COVERAGE_PLUGIN_MODULE_NAME)
+    config_parser.add_section(COVERAGE_PLUGIN_MODULE_NAME)
+    config_parser.set(
+        COVERAGE_PLUGIN_MODULE_NAME, "source_to_target_base", json.dumps(source_to_target_base)
+    )
+    config_parser.set(COVERAGE_PLUGIN_MODULE_NAME, "test_time", json.dumps(test_time))
+    config = StringIO()
+    config_parser.write(config)
+    return config.getvalue()
+
+
+class ReportType(Enum):
+    XML = "xml"
+    HTML = "html"
+
+
+class PytestCoverage(PythonToolBase):
+    options_scope = "pytest-coverage"
+    default_version = "coverage==5.0.3"
```

Comment:
Likely we want to use `>=5.0.3,<5.1` so that we automatically pull in bug fixes.

Reply:
The issue here is that we run two or three different process requests and coverage may throw errors if the version of coverage isn't exactly the same between all of them. Is there a way to ensure that we get exactly the same version across multiple pexes without pinning this?

Reply:
The only case where it won't be identical between them all is that half-way through the Pex invocations a new version of coverage is released to PyPI. With `coverage>=5.0.3,<5.1`, Pex/Pip will always resolve the newest release that is >= 5.03 and < 5.1. So, there is a possible race condition where one Pex invocation resolves 5.0.3 and then a couple seconds later 5.0.4 gets released to PyPI and the next invocation resolves that.

(Although, actually, I don't think that race condition can even occur due to V2 caching. We will cache the result of the EPR used to create the coverage Pex. So long as the inputs to the rule—such as the `--coverage-version` string—don't change, we won't try re-resolving  the Pex.)